packages/astro/src/assets/fonts/vite-plugin-fonts.ts

```diff
@@ -0,0 +1,237 @@
+				const resolvedOptions: unifont.ResolveFontOptions = {
+					weights: family.weights ?? DEFAULTS.weights,
+					styles: family.styles ?? DEFAULTS.styles,
+					subsets: family.subsets ?? DEFAULTS.subsets,
+					fallbacks: family.fallbacks ?? DEFAULTS.fallbacks,
+				};
+				// TODO: https://github.com/unjs/unifont/issues/108
+				const { fonts: fontsData, fallbacks } = await resolveFont(family.name, resolvedOptions, [
+					family.provider,
+				]);
+
+				// TODO: use fontaine if needed
+				const preloadData: PreloadData = [];
+				let css = '';
+				for (const data of fontsData) {
+					for (const source of data.src as unknown as Array<Record<string, string>>) {
+						const key = 'name' in source ? 'name' : 'url';
+						const hash = h64ToString(source[key]) + extname(source[key]);
+						const url = baseUrl + hash;
+						if (!collected.has(hash)) {
+							collected.set(hash, source[key]);
+							preloadData.push({ url, type: hash.split('.')[1] });
+						}
+						source[key] = url;
+					}
+					css += generateFontFace(family.name, data);
+				}
+				resolvedMap.set(family.name, { preloadData, css });
+			}
+		},
+		async configureServer(server) {
+			const logManager = createLogManager(logger);
+			// Base is taken into account by default
+			server.middlewares.use(URL_PREFIX, async (req, res, next) => {
+				if (!req.url) {
+					return next();
+				}
+				const hash = req.url.slice(1);
+				const url = collected?.get(hash);
+				if (!url) {
+					return next();
+				}
+				logManager.add(hash);
+				const { cached, data } = await cache!(hash, () =>
+					fetch(url)
+						.then((r) => r.arrayBuffer())
+						.then((arr) => Buffer.from(arr).toString()),
+				);
+				logManager.remove(hash, cached);
+
+				// TODO: add cache control back
+				// TODO: set content type and cache control manually
+				// const keys = ['cache-control', 'content-type', 'content-length'];
+				// const keys = ['content-type', 'content-length'];
+				// for (const key of keys) {
+				// 	const value = response.headers.get(key);
+				// 	if (value) {
+				// 		res.setHeader(key, value);
+				// 	}
+				// }
+				res.end(data);
+			});
+		},
+		resolveId(id) {
+			if (id === VIRTUAL_MODULE_ID) {
+				return RESOLVED_VIRTUAL_MODULE_ID;
+			}
+		},
+		load(id, opts) {
+			if (id === RESOLVED_VIRTUAL_MODULE_ID && opts?.ssr) {
+				return `
+				export const fontsData = new Map(${JSON.stringify(Array.from(resolvedMap?.entries() ?? []))})
+				`;
+			}
+		},
+		async buildEnd() {
+			resolvedMap = null;
+
+			if (sync) {
+				collected = null;
+				cache = null;
+				return;
+			}
+
+			const logManager = createLogManager(logger);
+			const dir = getBuildOutputDir(settings);
+			const fontsDir = new URL('.' + baseUrl, dir);
+			mkdirSync(fontsDir, { recursive: true });
+			await Promise.all(
+				Array.from(collected!.entries()).map(async ([hash, url]) => {
+					logManager.add(hash);
+					const { cached, data } = await cache!(hash, () =>
```

Comment:
Please add a comment that explains why `cache!` is safe. (essentially, we need to explain why `cache` is always set at this point). IMHO, it's safer to use an `if` block